src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,18 +1328,63 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            // There are a few different cases to handle when performing the write operation
+            // surrounding the internal buffer. Buffer flush operations can be issued asynchronously
+            // so the state of any async operation that touches the internal buffer must be tracked
+            // to ensure there are no data races at play.
+            //
+            //   1. No active async buffer op and numBytes fits in existing buffer:
+            //       Simply perform a memory copy operation and return synchronously.
+            //
+            //   2. Active async buffer op and numBytes <= _bufferSize
+            //       Since there's an active buffer operation, a new buffer needs to be allocated
+            //       to write to in order to ensure the existing buffer isn't modified mid-write.
+            //       The incoming write can then be copied into the new buffer.
+            //
+            //   3. No active async buffer and numBytes too large for buffer
+            //       If there's buffered data, issue a flush operation and then directly issue the
+            //       incoming write since it can't be buffered.
```

Comment:
The way this is phrased makes me as a reader think that the write is only issued if there's buffered data.  I'd suggest rephrasing it just slightly:
```
If there's buffered data, issue a flush operation.  Then, regardless, issue the incoming write, since it can't be buffered.
```